nifpga/session.py

```diff
@@ -390,8 +413,250 @@ def read(self):
+        super(_FxpRegister, self).__init__(session,
+                                           nifpga,
+                                           bitfile_register,
+                                           base_address_on_device,
+                                           read_func=nifpga["ReadArray%s" % DataType.U32],
+                                           write_func=nifpga["WriteArray%s" % DataType.U32])
+        self._word_length = bitfile_register.word_length
+        self._integer_word_length = bitfile_register.integer_word_length
+        self._overflow_enabled = bitfile_register.overflow
+        self._signed = bitfile_register.signed
+        self._delta = self._calculate_delta()
+        self._minimum = self._calculate_minimum()
+        self._maximum = self._calculate_maximum()
+        self._transfer_len = self._calculate_transfer_len()
+        self._ctype_type = self._ctype_type * self._transfer_len
+
+    def _calculate_delta(self):
+        """ The value persisted in the bitfile for delta is not always
+        correct, therefore we must calculate it manually.
+        """
+        return Decimal(2**(self._integer_word_length - self._word_length))
+
+    def _calculate_minimum(self):
+        """ The value persisted in the bitfile for the minimum value is
+        not always accurate, therefore we must calculate it manually.
+        """
+        if self._signed:
+            magnitude_bits = self._word_length - 1
+            return -1 * (2**(magnitude_bits) * self._delta)
+        else:
+            return 0
+
+    def _calculate_maximum(self):
+        """ The value persisted in the bitfile for the maximum value is not
+        always accurate, therefore we must calculate it manually.
+        """
+        if self._signed:
+            magnitude_bits = self._word_length - 1
+        else:
+            magnitude_bits = self._word_length
+        return (2**(magnitude_bits) - 1) * self._delta
+
+    def _calculate_transfer_len(self):
+        """ Fixed point values are transfered to the driver as an array of U32
+        The length is between 1 and 3 determined by the word length (includes
+        the signed bit) plus the include_overflow_status_enable bit.
+        """
+        bits_required = self._word_length
+        if self._overflow_enabled:
+            """ If overflow status is enabled we need an extra bit. """
+            bits_required += 1
+        if bits_required >= 32:
+            return int(ceil(bits_required / 32.0))
+        else:
+            return 1
+
+    def read(self):
+        """ Reads the fixed point representation from the register
+        offset, then converts it into a more native python experience.The
+        numeric value is returned as pythons Decimal. If the register has
+        the overflow status enabled the return value is a tuple, where the
+        first element is a bool for the overflow status, and the second is the
+        decimal value.
```

Comment:
Maybe:
""" Reads the fixed point value from the register and returns it as a Decimal.

    If the register..."""

Reply:
Done